### DISCUSSION TITLE
Using Random Forest classifier in batch mode

### DISCUSSION THREAD

**Discussion by vahuja4** [online-ml/river#554] [General]
Hello there,

I have a scenario in which I need to train a random forest classifier in the following manner:

Step1: Train on 20k datapoints just once
Step2: Train on 20 datapoints. This will repeat every week.

Can someone please tell me how this can be achieved using river-ml.

Sincerely,
Vishal

**Comment by MaxHalford** [comment]
Can you please tell us what you've tried? You should be able to do this in a straightforward manner by using the `learn_one` method. I'd be happy to help out, but I have no idea what you've already tried out.

Meta: the project is called River, not river-ml :)

**Reply by vahuja4** [reply]
Hi Max,
I have joined an organization recently and I am new to Deep Learning as well. The task at hand is to improve the accuracy of our CNNs using the audited data (humans correcting the predictions from the model). I don't want to retrain the CNN entirely with the updated data because that will be very expensive. So, here is what I thought of doing:
1. Extract the image representation for the entire dataset using the penultimate CNN layer
2. Train a random forest classifier using the image representation obtained in Step1.
3. Update only the random forest classifier using the audited data (once a week)

Apart from the time factor, there is a strong reason why I don't want to retrain the CNN using the audited data. There is only one CNN for multiple retailers. So, the audit data is coming in from all of them and I don't trust it to be of gold standard. If badly audited data were to be used to retrain the CNN, then it could adversely affect the predictions for all the other retailers. So, my plan is to have one random forest per retailer (trained in step2 above) and use it to make the actual predictions, and update it periodically only using the specific retailer's audit data. Two questions here:

1. Is this a good use-case for using river?
2. Does the adaptive random forest adapt to distribution shift in the production data?

**Reply by ElisonSherton** [reply]
Hi Max!

I am Vishal's colleague.

We have an attribute called dress-neckline which predicts one of the 11 different neckline types (like round, v, plunge etc.) The CNN model that we have trained is giving around 75% accuracy on the validation set and we chose to take the feature vectors from this model (512 dimensional representation for any given W x H x 3 image).

Here are some statistics for the training and validation sets

**Train Set Distribution**
![image](https://user-images.githubusercontent.com/45599762/113860663-e9b06300-97c3-11eb-9b86-bcd8401a4f1b.png)

**Valid Set Distribution**
![image](https://user-images.githubusercontent.com/45599762/113860750-02b91400-97c4-11eb-9a07-11a77764aa7e.png)

As we can observe it's a roughly equiprobable distribution with around 1800 datapoints per class in the train set and around 400 datapoints per class in the validation set.

I tried using river to progressively train an `AdaptiveRandomForestClassifier` with `learn_one` method as follows

```python
model = compose.Pipeline(
        preprocessing.StandardScaler(),
        ensemble.AdaptiveRandomForestClassifier(n_models=3, seed=42))

dataset = zip(xs, train_targets)

for x, y in tqdm(dataset, desc = f"Training..."):
    y_pred = model.predict_one(x)
    metric = metric.update(y, y_pred)
    model = model.learn_one(x, y) 
```

Each individual independent variable is a dictionary of 512 feature vectors
![image](https://user-images.githubusercontent.com/45599762/113861956-714aa180-97c5-11eb-9540-6c02aca7f52e.png)

When I print out the accuracy, I observe this
![image](https://user-images.githubusercontent.com/45599762/113862064-917a6080-97c5-11eb-89e8-6d5e6c4e8c47.png)
Which is really awesome! 

**BUT**

When I do the same thing without progressive learning by just doing a `predict_one` after finishing the training, this is what I obtain:

![image](https://user-images.githubusercontent.com/45599762/113862276-d2727500-97c5-11eb-8aa0-e2ab10f9d4e3.png)

Could you shed some light on why would there be such a high discrepancy in the evaluation of accuracy during progressive learning and at the end of learning? I will shortly provide a link to the notebook as well...

Thanks & Regards,
Vinayak.

**Reply by MaxHalford** [reply]
Hey guys. I believe a notebook would be very helpful.

**Reply by ElisonSherton** [reply]
Hi Max!

Yes, I can share the html file which was generated straight from the notebook here.

https://drive.google.com/file/d/19ZO20Um13N4LhkqkWr9chCu7TuJqJtRZ/view?usp=sharing

Let me know if you need any more idea about the problem at hand. 

Thanks & Regards,
Vinayak.

**Reply by MaxHalford** [reply]
I'm not able to see what's going wrong just by looking at the HTML. It definitely looks weird. Could you try using River's accuracy instead of scikit-learn's accuracy metric when you evaluate the model?

**Reply by ElisonSherton** [reply]
Hi Max!

I will check with my peers and get back to you if I can share the notebook and the data with you. But I tried River's accuracy here below and it is equivalent to scikit-learn's accuracy metric. You can have a look at the same below.

https://drive.google.com/file/d/1ePqlvPovmsVYmqN7XMN3UqTdnRMLfLvS/view?usp=sharing

Thanks & Regards,
Vinayak.

**Reply by ElisonSherton** [reply]
Hi Max,

Good Day!

I have consulted with Vishal and my other colleagues and I can give you the notebook and dataset here as below.

https://drive.google.com/drive/folders/13SIDHxDG9A3BC6rkRNjqKGJ75QOVqDHs?usp=sharing

There are two files of interest here:

1. *Using River ML For Adaptive Fitting a Random Forest.ipynb*: This is the notebook used for training
2. *dress_neckline_object.pkl*: This is the dataset which is utilized in the notebook for training.

You could ignore the two html files that I sent yesterday which are also present in this same folder. The environment details are as follows:

![image](https://user-images.githubusercontent.com/45599762/113966911-9682f280-984d-11eb-8ce6-42166e55c9dd.png)

I am using an ubuntu 18.04 LTS distribution as my Operating System. Hope this helps to investigate further for you.

Thanks & Regards,
Vinayak

**Reply by vahuja4** [reply]
Also Max, please look at these two questions of ours:
1. Is this a good use-case for using river?
2. Does the adaptive random forest adapt to distribution shift in the production data?

**Reply by MaxHalford** [reply]
Ah, I think I know. Can you try shuffling the training data? Online models tend to overfit on recent data if the data is ordered in a deceptive way. In this case I believe your training data is ordered by class.

> Is this a good use-case for using river?

To be honest I would be inclined to say no. I don't know how much data you have, but you can probably afford to retrain a batch model every week. 

> Does the adaptive random forest adapt to distribution shift in the production data?

Yes! That's one of the main benefits of using an online model.

**Reply by ElisonSherton** [reply]
Hey Max, 

Good Day!

Thanks for that suggestion of yours. Now that I run the code, I get a good concurrence between the results of train set when done dynamically and when done all at once as could be seen in the html code file below.

https://drive.google.com/file/d/10PTUs3MgO1WQpKS55JqL6guvNBlTTduy/view?usp=sharing

One question which I have Max is that let's say the incoming stream data is only of a single class; i.e. out of the 11 classes, I get 10 datapoints successively with label of 1 then will there be overfitting wrt a class?

Because what I can make out of this exercise is that when I curate the input in such a way that the same class' datapoints are iteratively feeded to the algorithm the accuracy as seen in the previous notebook is 98%. But it's generalizability goes for a toss in that case. Whereas when data is curated in such a way that the incoming datapoints are random or follow some kind of equiprobable distribution, then the generalizability is pretty decent.

Thanks & Regards,
Vinayak.

**Reply by MaxHalford** [reply]
Basically, you want to train the model with data ordered as it would be in production. The goal is to simulate as closely as possible a production scenario. It's likely that the data order is shuffled with respect to the classes in production so you should do the same thing for training.
